ethereum/spec/src/main/java/tech/pegasys/teku/spec/schemas/registry/AbstractSchemaProvider.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Consensys Software Inc., 2024
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.spec.schemas.registry;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.EnumMap;
+import java.util.Map;
+import tech.pegasys.teku.spec.SpecMilestone;
+import tech.pegasys.teku.spec.config.SpecConfig;
+import tech.pegasys.teku.spec.schemas.registry.SchemaTypes.SchemaId;
+
+abstract class AbstractSchemaProvider<T> implements SchemaProvider<T> {
+  private final Map<SpecMilestone, SpecMilestone> milestoneToEffectiveMilestone =
+      new EnumMap<>(SpecMilestone.class);
+  private final SchemaId<T> schemaId;
+
+  protected AbstractSchemaProvider(final SchemaId<T> schemaId) {
+    this.schemaId = schemaId;
+  }
+
+  protected void addMilestoneMapping(
+      final SpecMilestone milestone, final SpecMilestone untilMilestone) {
+    checkArgument(
+        untilMilestone.isGreaterThan(milestone),
+        "%s must be earlier than %s",
+        milestone,
+        untilMilestone);
+
+    SpecMilestone currentMilestone = untilMilestone;
+    while (!currentMilestone.equals(milestone)) {
+
+      checkIfAlreadyMapped(currentMilestone);
+
+      milestoneToEffectiveMilestone.put(currentMilestone, milestone);
+
+      currentMilestone = currentMilestone.getPreviousMilestone();
+    }
+
+    checkIfAlreadyMapped(currentMilestone);
+  }
+
+  private void checkIfAlreadyMapped(final SpecMilestone milestone) {
+    if (milestoneToEffectiveMilestone.containsKey(milestone)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Milestone %s is already mapped to %s",
+              milestone, milestoneToEffectiveMilestone.get(milestone)));
+    }
+  }
```

Comment:
I'd suggest separating the mapping overlapping check and make the add milestone mapping fails fast. I was confused when trying to understand the logic mainly because in some cases, we were adding the mapping and then failing.
What do you think if we use a `TreeMap` to hold a sorted `milestoneToEffectiveMilestone` and use the `floorEntry` and `ceilingEntry` to check the overlapping as follows:

```suggestion
  private final NavigableMap<SpecMilestone, SpecMilestone> milestoneToEffectiveMilestone =
      new TreeMap<>();
  private final SchemaId<T> schemaId;

  protected AbstractSchemaProvider(final SchemaId<T> schemaId) {
    this.schemaId = schemaId;
  }

  protected void addMilestoneMapping(
      final SpecMilestone milestone, final SpecMilestone untilMilestone) {
    checkArgument(
        untilMilestone.isGreaterThan(milestone),
        "%s must be earlier than %s",
        milestone,
        untilMilestone);

    checkOverlappingVersionMappings(milestone, untilMilestone);

    SpecMilestone currentMilestone = untilMilestone;
    while (!currentMilestone.equals(milestone)) {
      milestoneToEffectiveMilestone.put(currentMilestone, milestone);
      currentMilestone = currentMilestone.getPreviousMilestone();
    }
  }

  private void checkOverlappingVersionMappings(final SpecMilestone milestone, final SpecMilestone untilMilestone) {
    final Map.Entry<SpecMilestone, SpecMilestone> floorEntry =
        milestoneToEffectiveMilestone.floorEntry(untilMilestone);
    if (floorEntry != null && floorEntry.getValue().isGreaterThanOrEqualTo(milestone)) {
      throw new IllegalArgumentException(
          String.format(
              "Milestone %s is already mapped to %s",
              floorEntry.getKey(), getEffectiveMilestone(floorEntry.getValue())));
    }
    final Map.Entry<SpecMilestone, SpecMilestone> ceilingEntry =
        milestoneToEffectiveMilestone.ceilingEntry(milestone);
    if (ceilingEntry != null && ceilingEntry.getKey().isLessThanOrEqualTo(untilMilestone)) {
      throw new IllegalArgumentException(
          String.format(
              "Milestone %s is already mapped to %s",
              ceilingEntry.getKey(), getEffectiveMilestone(ceilingEntry.getValue())));
    }
  }
```

Reply:
I see what you mean, and your version could be indeed more readable. I'll check it out.